Sync typeshed

AlexWaygood opened the pull request:
Source commit:
https://github.com/python/typeshed/commit/d208d5a83394e9ad1d0dd5d567818c597dfaa28f

Changes in `ctypes` and `functools` caused merge conflicts with the cherry picks, and also broke the `ctypes` plugin (I hope the last commit in this PR fixes the plugin).

⚠️ Remember to use "rebase and merge" rather than "squash and merge" for this PR, so that we can easily update the commit hashes to cherry-pick in the misc/sync-typeshed.py script.

hauntsaninja commented:
Thanks for doing this! Looks great.

I was hopeful when we added the cherry picking that it would be rare that we'd have conflicts. This doesn't seem to have been borne out. I wonder if we'd be better served by just including the diffs in the repo and using `git apply`.

AlexWaygood commented:
> I was hopeful when we added the cherry picking that it would be rare that we'd have conflicts. This doesn't seem to have been borne out.

Yeah -- tbf, there has been a perhaps-unusual amount of churn in the stdlib stubs over recent months due to the "Let's add defaults everywhere!" project. Also, we now have 5 commits to cherry-pick each typeshed sync, which makes it much more likely there'll be a merge conflict somewhere than it was when we only had one commit to cherry-pick.

FWIW, I think the current system is still much better than the system we had before we introduced the automated workflow! I find the new script much easier to use, and it's nice to have an expectation of when the next typeshed sync is "due".

> I wonder if we'd be better served by just including the diffs in the repo and using `git apply`.

My git skills are fairly rudimentary, so I don't feel like I can comment on which will be better there really :p

JelleZijlstra commented:
I don't really like that we're doing these typeshed patches at all. It would be better if we could improve mypy so it works well with vanilla typeshed. However, that's a harder change to make.
